### PR TITLE
[SYCL][CUDA][MATRIX] joint_matrix_bmad implementation

### PR DESCRIPTION
**Pull request by JackAKirk**
cc @dkhaldi 

Implementation corresponding to the matrix extension proposal section "Bitwise Multiply and Add" in https://github.com/intel/llvm/pull/4695

Integration tests here: https://github.com/intel/llvm-test-suite/pull/760

### CODE REVIEW

File: sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp [hunk 1]
```diff
@@ -495,14 +562,59 @@ struct joint_matrix_mad_impl<
                                      get_layout_pair_id<LayoutA, LayoutB>(), 0);
         }
       }
-    } else if constexpr (std::is_same<T1, double>::value) {
+    } else if constexpr (M == 8 && N == 8 && K == 4) {
```
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp:565]: is this change related to bmad addition?

File: sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp [hunk 2]
```diff
@@ -339,6 +378,9 @@ struct joint_matrix_store_impl<
     } else if constexpr (std::is_same<T, double>::value) {
       __dmma_m8n8k4_st_c_f64(dst.get(), src.data, stride,
                              get_layout_id<Layout>());
+    } else if constexpr (std::is_same<T, int32_t>::value) {
+      __bmma_m8n8k128_st_c_i32(dst.get(), src.data, stride,
```
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp:382]: what is the difference between  __imma_m32n8k16_st_c_i32 and __bmma_m8n8k128_st_c_i32 other than the shape difference? 
Does "b" stand for binary here and it stores only results of bmad?
For my education, do you know how  K is so big in the bmad case (128)?

File: sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp [hunk 3]
```diff
@@ -235,6 +263,28 @@ struct joint_matrix_load_impl<
...
+        __dmma_m8n8k4_ld_c(res.data, src.get(), stride,
+                           get_layout_id<Layout>());
+      }
+    } else if constexpr (NumRows == 8 && NumCols == 128) {
+      int32_t *tileptr = reinterpret_cast<int32_t *>(src.get());
+      __bmma_m8n8k128_ld_a_b1(res.data, tileptr, stride,
+                              get_layout_id<Layout>());
+    } else if constexpr (NumRows == 128 && NumCols == 8) {
+      int32_t *tileptr = reinterpret_cast<int32_t *>(src.get());
+      __bmma_m8n8k128_ld_b_b1(res.data, tileptr, stride,
```
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp:286]: what types are supported in bmad? only double and i32?

File: sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp (hunk 1 above)
Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp:565]: No this is a superficial/non-important change that I made just for better consistency of the `if constexpr` statements in this function.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp (hunk 2 above)
Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp:382]: Yes "b" is for binary and it can only be used with the bmad case.  Since the shapes are different the two instructions load different number of registers/fragments for storing the C/Accumulator matrix type.  So it is important that there is the two distinct instructions in this case.

So k=128 because it refers to the number of bits in the K dimension which is equal to the number of elements in the K dimension for the bitwise case.  The number following k in these instructions always refers to a number of matrix elements, which may or may not equal the number of registers used to store these elements, depending on how the matrix elements are packed.  For the bmad cases we store 32 bit elements into a single int32_t register. For bf16 cases we store 2 elements into a single int32_t etc.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp (hunk 3 above)
Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp:286]: NVPTX bmad requires that matrix elements are stored in 32 bit untyped registers, int32_t is used here because when the NVPTX builtins for these functions were created int32_t register arguments were defined (uint32_t can also be used but there is no difference as far as NVPTX backend is concerned.).  As far as the user is concerned, I think that they should work with uint32_t for bmad cases as in https://github.com/intel/llvm-test-suite/pull/760.

double is not supported as a register storage type for bmad in NVPTX and I did not create a case for the user to use double with bmad.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp (hunk 2 above)
Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp:382]: In a related note it is worth noticing that there are two of these C load/store instructions e.g. `__mma_tf32_m16n16k8_ld_c` and `__hmma_m16n16k16_ld_c_f32` which when compiled down to the lower level assembly (SASS) for all compatible sm versions result in the exact same instructions!  We plan to take advantage of this to use the standard matrix extension interface for the tf32 type, which otherwise would require a new interface specialization for Nvidia backend.

Comment (dkhaldi) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp:382]: >>> So k=128 because it refers to the number of bits in the K dimension which is equal to the number of elements in the K dimension for the bitwise case. 

In the test case, this writing: joint_matrix<uint32_t, matrix_use::a, 8, 128,  matrix_layout::row_major> means
that 128 refers to number of uint32_t elements in a row of matrix A. How this is the same as number of bits as number of bits would be 128*32?

**Comment by JackAKirk** [comment]
Hi @dkhaldi 

If it is preferred for reviewing purposes I could add the temporary/initial fp19 implementation that uses uint32_t directly to this PR?  Hopefully the uint32_t fp19 should be a bit more straightforward to review compared to the bmad cases, since in the end we realized we can implement the fp19 cases in a way which is completely compliant with the existing matrix extension, whereas the bmad cases require a different interface.

Otherwise it is fine to put them up one at a time, I just thought it might make it easier to review them at once.

Thanks

**Comment by dkhaldi** [comment]
> If it is preferred for reviewing purposes I could add the temporary/initial fp19 implementation that uses uint32_t directly to this PR? Hopefully the uint32_t fp19 should be a bit more straightforward to review compared to the bmad cases, since in the end we realized we can implement the fp19 cases in a way which is completely compliant with the existing matrix extension, whereas the bmad cases require a different interface.

I think separate PRs is better.

**Comment by JackAKirk** [comment]
> I think separate PRs is better.

OK

File: sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp (hunk 2 above)
Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp:382]: I've now updated https://github.com/intel/llvm-test-suite/pull/760; I've renamed A/B as A_Packed/B_Packed, added comments making it clear that each array element of A_Packed contains 32 single-bit matrix elements.

Comment (dkhaldi) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp:382]: I think the implementation should change to have this in user code:
 joint_matrix<uint32_t, matrix_use::a, 8, 128/32, matrix_layout::row_major>;
bmad intrinsics (or the implementation) work on number of bits but users should not.
In the header file here, you should multiply by 32 to retrieve the number of bits to pass to the intrinsics rather than having divide by 32.
From the user perspective, they should operate on number of elements (not bits).

Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp:382]: You're probably right.  I've made the changes you describe. Thanks

File: sycl/test/check_device_code/matrix/matrix-nvptx-single-bit-test.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental::matrix;
+
+// M, N, K define the sizes of dimensions of the three matrix types (a, b,
+// accumulator) used per subgroup operation.
+constexpr int M = 8;   // number of rows of accumulator,
+                       // number of cols of b.
+constexpr int N = 8;   // number of cols of accumulator,
+                       // number of rows of a.
+constexpr int K = 128; // number of cols of a/number of rows of b.
```
Comment (dkhaldi) [inline sycl/test/check_device_code/matrix/matrix-nvptx-single-bit-test.cpp:16]: you missed to make the change here. 
K should be 4 here.
Can you please add a comment where you are making these changes?
Basically, saying that the underlying intrinsics are expecting a shape of K equals to number of total bits, not number of elements.

Comment (JackAKirk) [reply sycl/test/check_device_code/matrix/matrix-nvptx-single-bit-test.cpp]: Thanks I forgot that.  Fixed now, and I also updated the test so it will work with the legacy pass-manager.

I've added a more detailed comment describing Bitwise Dot Product and how this dictates the relation between the number of Array elements used for A/B arrays and the number of single-bit matrix elements that the A/B arrays represent.  I've also correspondingly updated the test in intel/llvm-test-suite and the tensor cores matrix extension PR https://github.com/intel/llvm/pull/4695.

File: sycl/test/check_device_code/matrix/matrix-nvptx-single-bit-test.cpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental::matrix;
+
+// M, N, (K * 32) define the sizes of dimensions of the three matrix types (a,
+// b, accumulator) used per subgroup operation.
+constexpr int M = 8; // number of rows of accumulator,
+                     // number of cols of b.
+constexpr int N = 8; // number of cols of accumulator,
+                     // number of rows of a.
+constexpr int K = 4; // number of cols of a/number of rows of b divided by 32
```
Comment (dkhaldi) [inline sycl/test/check_device_code/matrix/matrix-nvptx-single-bit-test.cpp:16]: >>> number of cols of a/number of rows of b divided by 32
should be:
number of bits in cols of a/number of bits in rows of B divided by 32.
If this is true, do we need to add the "divided by 32" in the code example.
I meant before to add the "multiplies by 32" in the implementation code to explain that this is how we get number of bits that exist in the intrinsics. But at the user level code, is this needed?

Comment (JackAKirk) [reply sycl/test/check_device_code/matrix/matrix-nvptx-single-bit-test.cpp:16]: Here K=4 is not the number of cols in a subgroup matrix: you have to multiply by 32, since K gives a dimension of the arrays A/B which hold the single-bit matrix elements in uint32_t storage type.  There are 32 single-bit matrix elements per uint32_t storage type.

I tried to describe the purpose of these bitwise matrix multiplications [here](https://github.com/intel/llvm/blob/ecb8e6bc1f75d7a3599c8b894d9b02a163b2842a/sycl/doc/extensions/Matrix/dpcpp-joint-matrix.asciidoc#motivation-for-bmad) without going into too much detail.  I added references for full details on the origins of the single-bit models and how they use the bitwise matrix multiplications within them. I do not find references to the usage of such "bitwise matrix multiplications" outside of such models (although of course this means there could be/will be in the future), but I think that
this functionality was introduced specifically with such a use case in mind.

It is important for such users to understand that each bit is considered an element of the matrix by `joint_matrix_bmad` (the matrix element is "quantized" to a single bit), which is why in the original implementation I set K = 128.  However as you pointed out this leads to lots of factors of 32 because we have to divide by 32 to get the number of uint32_t array elements that are used to store the matrix.  
In the current implementation it is nice that these factors are gone, but there should still be proper documentation describing the relationship between "K" and the actual number of (single-bit) matrix elements. Since this is experimental I think it is normal to expect that once people start using this there could be feedback suggesting small changes to the interface: I'm not sure whether the interface I originally set up that led to the factors of 32 or the one you suggested is preferable for the users, but I imagined that at this experimental stage it can (and I imagine most likely will!) be changed in some way in the future anyway.

I could add back the naming scheme A -> A_Packed, B-> B_Packed that I originally used when I switched to K=128 -> K=4 to make is clearer that I am calling "a" the matrix and "A_Packed" a packed array representation of the matrix?
Then I could also add some more detailed description in both the tests and the implementation?  I did not want to go into too much detail in tests/implementation because I thought that the proper place for such descriptions would be in the documentation of the extension? This is why I kept things concise here and did not mention details in the implementation.

**Comment by JackAKirk** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/760
